lib/types.nix

```diff
@@ -373,6 +373,14 @@ rec {
       merge = mergeEqualOption;
     };
 
+    hex = mkOptionType {
+      name = "hex";
+      description = "hexadecimal number";
+      descriptionClass = "noun";
+      check = x: str.check x && builtins.match "[0-9a-fA-F]+" x != null;
```

Comment:
Should we allow `0x` and/or `#`?

Reply:
I don't think so, at least for now. Can be added later if necessary

Reply:
`either` can convert failures into usable behavior, and functions like `coercedTo` can make a left match vs a right match in `either` observable, so allowing more _is_ a breaking change in non-trivial compositions of types.

Reply:
I'd imagine there's laws for types we could come up with to guard against this, e.g. for either:

> If `all aType.check aValues`, then
> ```nix
> (either bType aType).merge aValues
> == aType.merge aValues
> == (either aType bType).merge aValues
> ```

Reply:
Counterexample for the "laws":

```nix
aType = coercedTo hexStr (v: { base16 = v; }) attrs;
bType = coercedTo base64Str (v: { base64 = v; }) attrs;
# counterexample for "laws"
aValues = ["0000"];
```

Specifically,the definition is valid for both types, swapping the either runs a different coercion ie the inequalityj
```nix
(either bType aType).merge aValues
  != (either aType bType).merge aValues
```

Counterexample for allowing prefix later, reusing the types:

```nix
values = ["0x00"];

```


This causes `(either aType bType).merge` to return `{ base16 = ... }` instead the current `{ base64 = ... }` attrset.

Reply:
Oh yeah, I guess I haven't explained it fully, but that's exactly the intended point of this, because it would allow us to reason about which types should be allowed as `either` arguments and which shouldn't. And indeed your example sounds like something that shouldn't be allowed, so it seems to work. In practice we might introduce a typeclass/interface/trait hierarchy for the types and requiring `either` arguments to be of a certain typeclass/interface/trait. And given this requirement on either's argument, the law should hold.

Reply:
Laws about types are useless if they don't apply to all of them. I don't see use getting rid of `either` or `coercedTo` anytime soon.

Classifying types might be useful, but it doesn't change the behavior, and it doesn't make the discussed change a non-breaking one, except for a weaker definition of non-breaking using some of that new terminology.

It'd be good to have some theory to refer to in discussions like this, but we have much lower hanging documentation fruit to prioritize.

Reply:
Yeah sure, just felt like doing a bit of an exploration, let's not spend more cycles on this :)

(and I don't think this is blocking for this PR in any way)

Reply:
I agree. Hex data that's not necessarily a number (this type) is rarely prefixed anyway, so it looks like a change won't be necessary. No change implies no breaking change.

Resolving.